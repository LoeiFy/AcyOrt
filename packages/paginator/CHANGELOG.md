# @acyort/paginator

## 1.0.0

### Major Changes

- update acyort
