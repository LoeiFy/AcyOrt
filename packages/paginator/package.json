{
  "name": "@acyort/paginator",
  "version": "0.1.2",
  "description": "Pagination utilities",
  "main": "lib/index.js",
  "files": [
    "lib"
  ],
  "scripts": {
    "build": "rm -rf lib && tsc"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/acyortjs/acyort.git"
  },
  "keywords": [
    "pagination",
    "pages"
  ],
  "author": "LoeiFy <LoeiFy@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/acyortjs/acyort/issues"
  },
  "homepage": "https://github.com/acyortjs/acyort"
}
