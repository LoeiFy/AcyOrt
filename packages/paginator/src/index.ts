import { join } from 'path'
import { Params, PageCtx, PagingData } from './type'

function getPath(args: PageCtx) {
  const { current, prefix, baseUrl } = args
  if (current === 1) {
    return baseUrl
  }
  return join(baseUrl, prefix, current.toString(), '/')
}

function getPrev(args: PageCtx) {
  const { current, prefix, baseUrl } = args

  if (current === 1) {
    return ''
  }
  if (current === 2) {
    return join(baseUrl, '/')
  }
  return join(baseUrl, prefix, (current - 1).toString(), '/')
}

function getNext(args: PageCtx) {
  const { current, prefix, baseUrl } = args
  return join(baseUrl, prefix, (current + 1).toString(), '/')
}

export default (args: Params): PagingData[] => {
  const {
    baseUrl,
    perpage,
    data,
    prefix = 'page',
  } = args

  if (perpage === 0 || data.length <= perpage) {
    return [{
      baseUrl,
      prev: '',
      next: '',
      data,
      fullPath: getPath({ baseUrl, current: 1, prefix }),
      current: 1,
      total: 1,
    }]
  }

  const result: PagingData[] = []
  const total = Math.ceil(data.length / perpage)

  let page = 1

  for (let i = 0; i < data.length; i += perpage) {
    result.push({
      baseUrl,
      prev: getPrev({ baseUrl, current: page, prefix }),
      next: getNext({ baseUrl, current: page, prefix }),
      data: data.slice(i, i + perpage),
      fullPath: getPath({ baseUrl, current: page, prefix }),
      current: page,
      total,
    })

    if (page === total) {
      result[page - 1].next = ''
    }

    page += 1
  }

  return result
}
