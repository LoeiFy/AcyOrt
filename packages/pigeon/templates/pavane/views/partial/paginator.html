<div class="paginator">
  <div>
    <a{% if page.prev %} href="{{ _url(page.prev) }}"{% endif %}>{{ __('previous') }}</a>
    <a{% if page.next %} href="{{ _url(page.next) }}"{% endif %}>{{ __('next') }}</a>
  </div>
  <span>{{ page.current }} / {{ page.total }}</span>
</div>
