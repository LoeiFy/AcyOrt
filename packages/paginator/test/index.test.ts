import paginator from '../src'

it('no pages', () => {
  const res = paginator({
    baseUrl: '/xxx/yyy',
    perpage: 0,
    data: [1, 2, 3],
  })
  expect(res).toMatchSnapshot()
})

it('pages', () => {
  const res = paginator({
    baseUrl: '/',
    perpage: 2,
    data: [1, 2, 3, 4, 5],
  })
  expect(res).toMatchSnapshot()
})

it('prefix', () => {
  const res = paginator({
    baseUrl: '/',
    perpage: 3,
    data: [1, 2, 3, 4, 5],
    prefix: 'nav',
  })
  expect(res).toMatchSnapshot()
})
