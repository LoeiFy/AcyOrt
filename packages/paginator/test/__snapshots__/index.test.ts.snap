// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no pages 1`] = `
[
  {
    "baseUrl": "/xxx/yyy",
    "current": 1,
    "data": [
      1,
      2,
      3,
    ],
    "fullPath": "/xxx/yyy",
    "next": "",
    "prev": "",
    "total": 1,
  },
]
`;

exports[`pages 1`] = `
[
  {
    "baseUrl": "/",
    "current": 1,
    "data": [
      1,
      2,
    ],
    "fullPath": "/",
    "next": "/page/2/",
    "prev": "",
    "total": 3,
  },
  {
    "baseUrl": "/",
    "current": 2,
    "data": [
      3,
      4,
    ],
    "fullPath": "/page/2/",
    "next": "/page/3/",
    "prev": "/",
    "total": 3,
  },
  {
    "baseUrl": "/",
    "current": 3,
    "data": [
      5,
    ],
    "fullPath": "/page/3/",
    "next": "",
    "prev": "/page/2/",
    "total": 3,
  },
]
`;

exports[`prefix 1`] = `
[
  {
    "baseUrl": "/",
    "current": 1,
    "data": [
      1,
      2,
      3,
    ],
    "fullPath": "/",
    "next": "/nav/2/",
    "prev": "",
    "total": 2,
  },
  {
    "baseUrl": "/",
    "current": 2,
    "data": [
      4,
      5,
    ],
    "fullPath": "/nav/2/",
    "next": "",
    "prev": "/",
    "total": 2,
  },
]
`;
